aptible/resource_app.go

```diff
@@ -52,47 +54,217 @@ func resourceApp() *schema.Resource {
+				Description: "Cooldown period in seconds after a release event.",
+			},
+			"mem_cpu_ratio_r_threshold": {
+				Type:     schema.TypeFloat,
+				Default:  4.0,
+				Optional: true,
+			},
+			"mem_cpu_ratio_c_threshold": {
+				Type:     schema.TypeFloat,
+				Default:  2.0,
+				Optional: true,
+			},
+			"mem_scale_up_threshold": {
+				Type:        schema.TypeFloat,
+				Optional:    true,
+				Default:     0.9,
+				Description: "The memory usage threshold for scaling up.",
+			},
+			"mem_scale_down_threshold": {
+				Type:        schema.TypeFloat,
+				Optional:    true,
+				Default:     0.75,
+				Description: "The memory usage threshold for scaling down.",
+			},
+			"minimum_memory": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     2048,
+				Description: "The minimum memory allocation in MB.",
+			},
+			"maximum_memory": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "The maximum memory allocation in MB.",
+			},
+			"min_cpu_threshold": {
+				Type:        schema.TypeFloat,
+				Optional:    true,
+				Description: "The minimum CPU utilization threshold for scaling.",
+			},
+			"max_cpu_threshold": {
+				Type:        schema.TypeFloat,
+				Optional:    true,
+				Description: "The maximum CPU utilization threshold for scaling.",
+			},
+			"min_containers": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "The minimum number of containers for scaling.",
+			},
+			"max_containers": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "The maximum number of containers for scaling.",
+			},
+			"scale_up_step": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     1,
+				Description: "The number of containers to add in each scale-up event.",
+			},
+			"scale_down_step": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     1,
+				Description: "The number of containers to remove in each scale-down event.",
 			},
 		},
 	}
 }
 
+func validateServiceSizingPolicy(ctx context.Context, d *schema.ResourceDiff, _ interface{}) error {
+	services := d.Get("service").(*schema.Set).List()
+	for _, service := range services {
+		serviceMap := service.(map[string]interface{})
+		if policy, ok := serviceMap["service_sizing_policy"]; ok {
+			policies := policy.(*schema.Set).List() // should only be one, but yes it's a Set
+			if len(policies) == 1 && policies[0] != nil {
+				policyMap := policies[0].(map[string]interface{})
+				autoscalingType := policyMap["autoscaling_type"].(string)
+				attrsToCheck := []string{
+					"min_containers",
+					"max_containers",
+					"min_cpu_threshold",
+					"max_cpu_threshold",
+				}
+				if autoscalingType == "horizontal" {
+					for _, attr := range attrsToCheck {
+						if val, ok := policyMap[attr]; !ok || val == nil || val == 0 {
+							return fmt.Errorf("%s is required when autoscaling_type is set to 'horizontal'", attr)
```

Comment:
These errors are being returned in the services loop.  So when there's an error here it basically bails out and throws an error in terraform.  I assume that is desired since it is likely user-error, just wanted to call it out and make sure this was intentional.